File: src/MdtUniversale.java
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

public class MdtUniversale /* extends HashMap<Integer,String> */ {

	/**
	 * 
	 */
	private static final long serialVersionUID = 5171164049513756757L;
	public HashMap<Integer, String> file;
	private List<HashMap<String, String>> nastro1;
	private String statoAttuale;
	private int testa;
	private String stringaInserita;
	private int maxComputazioni;
	private LinkedList<String> nastroStringa;
	private BinarieTransizioni bt;

	public MdtUniversale() {
		this.file = (new FakeFile()).getFile();
		bt = new BinarieTransizioni();

		this.nastro1 = bt.convertiTransizioniInBinario(bt.leggi_transizioni(file));
		if (this.nastro1.size() == 0) {
			System.out.println("Non è presente alcuna descrizione nel file");
		}
		this.statoAttuale = this.nastro1.get(0).get("stato_partenza");
		this.testa = 0;
		this.maxComputazioni = 100000;
		this.stringaInserita = " ";
	}

	public HashMap<String, String> calcoloConfigurazioneSuccessiva(String stato, String simbolo) {
		int i = 0;
		boolean trovato = false;
		HashMap<String, String> configurazione = new HashMap<String, String>();
		while ((i < nastro1.size()) && (trovato == false)) {
			if(this.nastro1.get(i).get("simbolo_letto").equals(this.nastroStringa.get(this.testa)) && this.statoAttuale.equals(this.nastro1.get(i).get("stato_partenza"))) {
				configurazione.put("simbolo", this.nastro1.get(i).get("simbolo_scritto"));
				configurazione.put("stato", this.nastro1.get(i).get("stato_arrivo"));
				configurazione.put("movimento", this.nastro1.get(i).get("movimento"));
				
				trovato = true;
			}
			i = i+1;
		}
		return configurazione;
	}

	public void computaStringa(String input) throws Exception {
		if (this.testa == 0 || this.statoAttuale.equals(this.nastro1.get(0).get("stato_partenza"))) {
			this.reset();
		}

		this.stringaInserita = input;
		this.nastroStringa = bt.stringaALista(bt.convertiStringaInBinario(input));
		int comp = this.computa();
		this.stampaRisultati(comp);
	}
	
	private int computa() throws Exception {
		boolean stop = false;
		int comp = 0;
		
		while(!stop) {
			if(this.statoAttuale.equals(bt.getAccetta())) {
				stop = true;
			}
			
			HashMap<String, String> configurazione = this.calcoloConfigurazioneSuccessiva(this.statoAttuale, this.nastroStringa.get(this.testa));
			
			if(configurazione.size() == 0) {
				stop = true;
			} else {
				this.statoAttuale = configurazione.get("stato");
				this.nastroStringa.set(this.testa, configurazione.get("simbolo"));

				// calcolo il movimento da fare sul nastro della stringa
				String movimentoLetto = configurazione.get("movimento");
				
				if(movimentoLetto.equals(bt.getSinistra())) {
					if(this.testa == 0) {
						this.nastroStringa.add(0, bt.getVuoto());
					} else {
						this.testa--;
					}
				} else if(movimentoLetto.equals(bt.getDestra())) {
					if(this.testa + 1 == this.nastroStringa.size()) {
						this.nastroStringa.add(bt.getVuoto());
						this.testa++;
					} else {
						this.testa++;
					}
				} else {
					throw new Exception("MdtUniversale.computa(): Qualcosa è andato storto");
				}
			}
			
			comp++;
			
			if(comp == this.maxComputazioni) {
				stop = true;
			}
		}
		return comp;
	}
	
	private void reset() {
		this.stringaInserita = " ";
		this.testa = 0;
		this.nastroStringa = new LinkedList<String>();
		this.statoAttuale = this.nastro1.get(0).get("stato_partenza");
	}

	public void stampaRisultati(int nComputazioni) {
		System.out.println("Macchina di Turing in input:");
		bt.stampaListaTransizioni(bt.leggi_transizioni(this.file));

		System.out.println("Macchina di Turing in binario:");
		bt.stampaListaTransizioni(this.nastro1);

		System.out.println("Stringa in input: " + this.stringaInserita);
		System.out.println(
				"Stringa in binario computata dalla macchina: " + bt.convertiStringaInBinario(stringaInserita));

		// stato_stampato = self.stato_attuale
		this.stampaNastro();

		if (nComputazioni == this.maxComputazioni)
			System.out.println("La macchina e' andata in loop");
		else
			System.out.println("Numero di computazioni: " + nComputazioni);

		if (this.statoAttuale.equals(bt.getAccetta()))
			System.out.println("La stringa e' stata accettata");
		else
			System.out.println("La stringa e' stata rifiutata");
	}

	public void stampaNastro() {
		LinkedList<String> nastroStampato = new LinkedList<String>();
		for(int i = 0; i< nastroStringa.size(); i++) {
			if(!nastroStringa.get(i).equals(bt.getVuoto())) {
				nastroStampato.add(nastroStringa.get(i));
			}
		}
		
		String stringaStampata = "";
		for(String s : nastroStampato) {
			String sASCII = bt.convertiBinarioInStringa(s);
			stringaStampata += sASCII;
			stringaStampata += " ";
		}
		System.out.println(stringaStampata);
	}
	
	public String toString() {
		return file.get(1) + "\n" + file.get(2) + "\n" + file.get(3) + "\n" + file.get(4) + "\n" + file.get(5) + "\n"
				+ file.get(6) + "\n" + file.get(7);
	}

}

File: src/mdt_universale.py
from TransizioniBinarie import *


class mdt_universale:

    def __init__(self, file_input_transizioni):

        self.__file = file_input_transizioni

        self.nastro1 = convert_binario_transizioni(
            leggi_transizioni(file_input_transizioni))

        if len(self.nastro1) == 0:
            raise ValueError("Non e' presente alcuna descrizione nel file")

        self.stato_attuale = self.nastro1[0]["stato_partenza"]
        self.testina = 0

        self.__stringa_inserita = " "
        self.__MAX_COMPUTAZIONI = 100000

    def calcola_prossima_configurazione(self, stato, simbolo):

        i = 0
        trovato = False
        configurazione = dict()

        while i < len(self.nastro1) and trovato == False:

            if self.nastro1[i]["simbolo_letto"] == self.nastro_stringa[self.testina] and self.stato_attuale == self.nastro1[i]["stato_partenza"]:

                configurazione["simbolo"] = self.nastro1[i]["simbolo_scritto"]
                configurazione["stato"] = self.nastro1[i]["stato_arrivo"]
                configurazione["movimento"] = self.nastro1[i]["movimento"]

                trovato = True

            i = i + 1

        return configurazione

    def __computa(self):

        stop = False
        comp = 0

        while stop == False:

            if self.stato_attuale == accept:
                stop = True

            configurazione = self.calcola_prossima_configurazione(
                self.stato_attuale, self.nastro_stringa[self.testina])

            if len(configurazione) == 0:
                stop = True

            else:
                self.stato_attuale = configurazione['stato']
                self.nastro_stringa[self.testina] = configurazione['simbolo']

                # calcolo il movimento da fare sul njastro della stringa
                movimento_letto = configurazione['movimento']

                if movimento_letto == left:

                    if self.testina == 0:
                        self.nastro_stringa.insert(0, empty)

                    else:
                        self.testina = self.testina - 1

                elif movimento_letto == right:

                    if self.testina + 1 == len(self.nastro_stringa):
                        self.nastro_stringa.append(empty)
                        self.testina = self.testina + 1

                    else:
                        self.testina = self.testina + 1

                else:
                    raise ValueError("Qualcosa è andato storto")

            comp = comp + 1

            if comp == self.__MAX_COMPUTAZIONI:
                stop = True

        return comp

    # metodo privato
    def __reset(self):
        self.__stringa_inserita = " "
        self.testina = 0
        self.nastro_stringa = list()
        self.stato_attuale = self.nastro1[0]["stato_partenza"]

    def computa_stringa(self, stringa_input):

        if self.testina == 0 or self.stato_attuale == self.nastro1[0]["stato_partenza"]:
            self.__reset()

        self.__stringa_inserita = stringa_input
        self.nastro_stringa = stringa_a_lista(
            convert_binario_carattere(stringa_input))
        comp = self.__computa()
        self.__stampa_risultati(comp)

    def __stampa_nastro(self):
        nastro_stampato = self.nastro_stringa
        nastro_stampato = [i for i in nastro_stampato if i != empty]

        stringa_stampata = ""

        for s in nastro_stampato:
            s_ascii = convert_ascii_binario(s)
            stringa_stampata += s_ascii
            stringa_stampata += ' '

        print(stringa_stampata)

    def __stampa_risultati(self, n_computazioni):

        print("Macchina di Turing in input: \n")
        print_lista_transizioni(leggi_transizioni(self.__file))

        print("Macchina di Turing in binario: \n")
        print_lista_transizioni(self.nastro1)

        print("Stringa in input: %s" % self.__stringa_inserita)
        print("Stringa in binario computata dalla macchina: %s" % convert_binario_carattere(self.__stringa_inserita))

        stato_stampato = self.stato_attuale
        self.__stampa_nastro()

        if n_computazioni == self.__MAX_COMPUTAZIONI:
            print("La macchina e' andata in loop")

        else:
            print("Numero di computazioni: %d" % n_computazioni)

        if self.stato_attuale == accept:
            print("La stringa e' stata accettata")

        else:
            print("La stringa e' stata rifiutata")
